### ISSUE TITLE
Cypher ignores WHERE part

### ISSUE THREAD

**Issue by riediger** [neo4j/neo4j#12618]
I suspect a bug in the Cypher planner/optimizer. A whole part of the query mentioned below isn't included in the plan, resulting in an empty result.

I've tried to execute this query on a database that I can't provide here due to intellectual property issues. However, the bug is also reproducible on an empty database, an anonymized graph is provided as cypher export.
 
**Neo4j Version:** Neo4j Server version: 4.1.1 (enterprise), single machine
**Operating System:** MacOS 10.15.7
**API:** Neo4j Browser version: 4.1.4

### Steps to reproduce

1. Create an empty database
2. Connect to that database using Neo4j Browser
3. Unzip `data.zip` (attached, contains example graph, queries, expected output, query plan)

[data.zip](https://github.com/neo4j/neo4j/files/5474945/data.zip)

4. Create nodes and relations by executing statements from `graph.cypher
5. Execute `query.cypher`

```
MATCH (td1 :TailoringDecision) -[:TAILORS]-> (re2 :RequirementSet) -[:INCLUDES*1..]-> (re1)
WHERE NOT EXISTS {
  MATCH (td2 :TailoringDecision)
  WHERE ((td2) -[:TAILORS]-> (re1))
    OR (EXISTS {
      MATCH (td2) -[:TAILORS]-> (re3 :RequirementSet) <-[:INCLUDES*1..]- (re2)
      WHERE (re3) -[:INCLUDES]-> (re1)
    })
}
RETURN td1, re2.userId, collect(re1.userId)
```

### Expected behavior
- The output should look like `expected-result.json` (calculated using `alternative-query.cypher`, an inefficient work-around)
- The query plan (created by prepending `EXPLAIN`) should contain operators to evaluate the nested `WHERE` predicates.

### Actual behavior
- Empty output.
- No operators are included in the query plan to compute the `WHERE` predicates of the nested `MATCH`, hence there exists a match for `td2`, hence the result is empty.

**Comment by mnd999** [comment]
Thanks for the report. We’ll look into it and come back when we have a solution.

**Comment by Lojjs** [comment]
@riediger This was indeed a bug in the Cypher planner, which we have been able to reproduce on Neo4j 4.1. It is already fixed in Neo4j 4.2, which was released this week, so one option for you is to upgrade to 4.2. Meanwhile, we are looking into backporting the bug fix to 4.1 and including it in an upcoming maintenance release.

Best regards Louise, Neo4j Cypher team

**Comment by riediger** [comment]
Great, thank you 👍 
Best wishes: Volker
